File: core/frontend.js
// +build js,wasm

package main

import (
	"encoding/json"
	"syscall/js"
)

type genericFunction func(map[string]interface{}) map[string]interface{}

func genericWrapper(args []js.Value, function genericFunction) interface{} {
	arguments := make(map[string]interface{})
	if fail := json.Unmarshal([]byte(args[0].String()), &arguments); fail == nil {
		if result, fail := json.Marshal(function(arguments)); fail == nil {
			return string(result)
		} else {
			return nil
		}
	} else {
		return nil
	}
}

var lut = map[string]genericFunction{
	"feature01":  feature01,
	"feature02":  feature02,
	"feature03":  feature03,
	"feature04a": feature04A,
	"feature04b": feature04B,
	"feature05":  feature05,
	"feature06":  feature06,
	"feature07":  feature07,
	"feature08a": feature08A,
	"feature08b": feature08B,
	"feature09a": feature09A,
	"feature09b": feature09B,
	"feature10":  feature10,
	"feature11":  feature11,
	"feature12":  feature12,
	"feature13":  feature13,
	"feature14":  feature14,
	"feature15":  feature15,
	"feature16":  feature16,
	"feature17":  feature17,
	"feature18":  feature18,
	"feature19":  feature19,
	"feature20":  feature20,
	"feature21":  feature21,
	"feature22":  feature22,
	"feature23":  feature23,
	"feature24":  feature24,
	"feature25":  feature25,
	"feature26":  feature26,
	"feature27":  feature27,
	"feature28":  feature28,
	"feature29":  feature29,
	"feature30":  feature30,
	"feature31":  feature31,
	"feature32":  feature32,
	"feature33":  feature33,
	"feature34":  feature34,
	"feature35":  feature35,
	"feature39":  feature39,
	"feature40":  feature40,
	"feature41":  feature41,
	"feature43":  feature43,
	"feature44":  feature44,
	"feature45":  feature45,
	"feature46":  feature46,
	"feature47":  feature47,
	"feature48":  feature48,
	"feature49":  feature49,
	"feature50":  feature50,
	"feature51":  feature51,
	"feature52":  feature52,
	"feature53":  feature53,
	"feature54":  feature54,
	"feature55":  feature55,
	"feature56":  feature56,
	"feature57":  feature57,
}

func main() {
	window := js.Global().Get("window")
	for name := range lut {
		function, _ := lut[name]
		window.Set(name, js.FuncOf(func(this js.Value, args []js.Value) interface{} {
			return genericWrapper(args, function)
		}))
	}
	select {}
}

File: core/frontend.go
// +build js,wasm

package main

import (
	"fmt"
	"syscall/js"
)

func arrayToJS(array []float64) interface{} {
	result := make([]interface{}, 0)
	for _, value := range array {
		result = append(result, value)
	}
	return result
}

func getValue(object js.Value, field string) (js.Value, error) {
	value := object.Get(field)
	if value.IsUndefined() {
		return value, fmt.Errorf("Value is undefined")
	}
	if value.IsNull() {
		return value, fmt.Errorf("Value is null")
	}
	if value.IsNaN() {
		return value, fmt.Errorf("Value is NaN")
	}
	return value, nil
}

func feature01Wrapped(this js.Value, args []js.Value) interface{} {
	var dd, rr, ss, ee, tt float64
	if value, fail := getValue(args[0], "dd"); fail == nil {
		dd = value.Float()
	} else {
		return nil
	}
	if value, fail := getValue(args[0], "rr"); fail == nil {
		rr = value.Float()
	} else {
		return nil
	}
	if value, fail := getValue(args[0], "ss"); fail == nil {
		ss = value.Float()
	} else {
		return nil
	}
	if value, fail := getValue(args[0], "ee"); fail == nil {
		ee = value.Float()
	} else {
		return nil
	}
	if value, fail := getValue(args[0], "tt"); fail == nil {
		tt = value.Float()
	} else {
		return nil
	}
	s, ll1, ll2, p, a := feature01(dd, rr, ss, ee, tt)
	result := make(map[string]interface{})
	result["s"] = s
	result["ll1"] = ll1
	result["ll2"] = ll2
	result["p"] = arrayToJS(p)
	result["a"] = a
	return result
}

func main() {
	window := js.Global().Get("window")
	window.Set("feature01", js.FuncOf(feature01Wrapped))
	select {}
}
